### PR TITLE
refactor(server): access permissions

### PR DESCRIPTION
**Pull request by jrasm91**
Create an `AccessCore` where all asset, album, etc. access checks can be consolidated. Eventually, hopefully, this logic can be enhanced with additional modifiers (additional share settings set on the user, album, asset, api key, etc.)

### CODE REVIEW

File: server/src/immich/api-v1/asset/asset.service.ts [hunk 1]
```diff
@@ -421,15 +413,14 @@ export class AssetService {
...
-    await this.checkAssetsAccess(authUser, dto.ids, true);
-
     const deleteQueue: Array<string | null> = [];
     const result: DeleteAssetResponseDto[] = [];
 
     const ids = dto.ids.slice();
     for (const id of ids) {
       const asset = await this._assetRepository.get(id);
-      if (!asset) {
+      const hasAccess = asset && (await this.access.hasPermission(authUser, Permission.ASSET_DELETE, id));
```
Comment (brighteyed) [inline server/src/immich/api-v1/asset/asset.service.ts:422]: Not a big deal, but maybe this code looks more natural:
```
      const hasAccess = await this.access.hasPermission(authUser, Permission.ASSET_DELETE, id);
      if (!hasAccess) {
        result.push({ id, status: DeleteAssetStatusEnum.FAILED });
        continue;
      }
      const asset = await this._assetRepository.get(id);
      if (!asset) {
        result.push({ id, status: DeleteAssetStatusEnum.FAILED });
        continue;
      }
```

File: server/src/immich/api-v1/asset/asset.service.ts [hunk 2]
```diff
@@ -208,32 +211,20 @@ export class AssetService {
   }
 
   public async getAllAssets(authUser: AuthUserDto, dto: AssetSearchDto): Promise<AssetResponseDto[]> {
-    if (dto.userId && dto.userId !== authUser.id) {
-      await this.checkUserAccess(authUser, dto.userId);
-    }
+    await this.access.requirePermission(authUser, Permission.LIBRARY_READ, dto.userId || authUser.id);
```
Comment (brighteyed) [inline server/src/immich/api-v1/asset/asset.service.ts:214]: I'd prefer to extract variable `const userId = dto.userId || authUser.id` like in `getAssetByTimeBucket` and `getAssetCountByTimeBucket`

**Review by brighteyed — APPROVED** [review]
I like this solution! It really simplifies permission-checking for services

**Comment by alextran1502** [comment]
## Issue 1
Not sure if this is related to this PR. In public shared links, drag and drop to upload returns `forbidden` status while clicking on the upload button to choose assets from the file explorer works.

## Issue 2
Cosmetic - on the web, with shared link disabled download, while multi selection enabled, the app bar showed the download button, although it doesn't work.

Let me know if you want to address those two issues in this PR or we can do it in a subsequent one

File: server/src/domain/access/access.core.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        return await this.repository.asset.hasSharedLinkAccess(sharedLinkId, id);
+
+      case Permission.ASSET_DOWNLOAD:
+        return !!authUser.isAllowDownload && (await this.repository.asset.hasSharedLinkAccess(sharedLinkId, id));
+
+      case Permission.ASSET_SHARE:
+        // TODO: fix this to not use authUser.id for shared link access control
+        return this.repository.asset.hasOwnerAccess(authUser.id, id);
+
+      // case Permission.ALBUM_READ:
```
Comment (alextran1502) [inline server/src/domain/access/access.core.ts:71]: Is this for debugging, should we keep or remove these lines?

File: server/src/domain/access/access.core.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
+import { BadRequestException } from '@nestjs/common';
+import { AuthUserDto } from '../auth';
+import { IAccessRepository } from './access.repository';
+
+export enum Permission {
+  // ASSET_CREATE = 'asset.create',
```
Comment (alextran1502) [inline server/src/domain/access/access.core.ts:6]: Are we planning to use these commented-out permission later?

Comment (jrasm91) [reply server/src/domain/access/access.core.ts:6]: Yes, I added in everything when i was designing it, but decided not to add permission checks where none previously existed in this PR. In a subsequent one these will be enabled and used.

File: server/src/domain/access/access.core.ts (hunk 3 above)
Comment (jrasm91) [reply server/src/domain/access/access.core.ts:71]: I can remove them if you want. I believe all the commented out lines will be used in the future as more things migrate to this permission model.

**Comment by jrasm91** [comment]
I'd guess issue one and two are pre-existing, and if so they can be fixed in a separate PR. If they are new bugs I'll fix them obviously. Those are good finds though and I can probably take a look at fixing them soon.

**Comment by jrasm91** [comment]
I confirmed issue one and two exist in the demo site today, so they're not introduced with this PR.

**Pull request merged**